uefi/firmwarefile.go

```diff
@@ -115,13 +139,95 @@ type FirmwareFile struct {
 	DataOffset  uint64
 }
 
-// Assemble assembles the Firmware File from the binary
+// Assemble assembles the Firmware File
 func (f *FirmwareFile) Assemble() ([]byte, error) {
 	var err error
-	f.buf, err = ioutil.ReadFile(f.ExtractPath)
+
+	fh := &f.Header
+	if _, ok := supportedFiles[fh.Type]; !ok {
+		// we don't support this file type, just return the raw buffer.
+		f.buf, err = ioutil.ReadFile(f.ExtractPath)
+		if err != nil {
+			return nil, err
+		}
+		return f.buf, nil
+	}
+
+	// Otherwise, we reconstruct the entire file from the sections and the
+	// file header using data from the JSON. This means that some JSON values
+	// will now be respected, including GUID changes. However file lengths and
+	// checksums will be recalculated.
+
+	// Assemble all sections so we know the final file size. We need to do this
+	// to know if we need to use the extended header.
+	fileData := []byte{}
+	dLen := uint64(0)
```

Comment:
What is the purpose of `dLen`? Won't it always equal `len(fileData)` after any operation?